diff --git a/haskell/cabal.bzl b/haskell/cabal.bzl
index b8b68074..aac1435c 100644
--- a/haskell/cabal.bzl
+++ b/haskell/cabal.bzl
@@ -100,7 +100,12 @@ def _prepare_cabal_inputs(hs, cc, posix, dep_info, cc_info, component, package_i
 
     (ghci_extra_libs, env) = get_ghci_extra_libs(hs, posix, cc_info)
     env.update(**hs.env)
-    env["PATH"] = join_path_list(hs, _binary_paths(tool_inputs) + posix.paths)
+    if hs.label.workspace_name == "proto3_suite":
+        cc_path = paths.dirname(cc.tools.cpp)
+        print("cc_path", cc_path)
+        env["PATH"] = join_path_list(hs, _binary_paths(tool_inputs) + posix.paths + [cc_path])
+    else:
+        env["PATH"] = join_path_list(hs, _binary_paths(tool_inputs) + posix.paths)
     if hs.toolchain.is_darwin:
         env["SDKROOT"] = "macosx"  # See haskell/private/actions/link.bzl
 
@@ -122,6 +127,28 @@ def _prepare_cabal_inputs(hs, cc, posix, dep_info, cc_info, component, package_i
     ]
     args.add_all([component, package_id, setup, cabal.dirname, package_database.dirname])
     args.add("--flags=" + " ".join(flags))
+    if hs.label.workspace_name == "proto3_suite":
+        execroot = "C:/users/admin/_bazel_admin/oaoes6io/execroot/com_github_digital_asset_daml/"
+        args.add_all([
+            # GHC uses C compiler for assemly, linking and preprocessing as well.
+            "-pgma",
+            execroot + cc.tools.cc,
+            "-pgmc",
+            execroot + cc.tools.cc,
+            "-pgml",
+            execroot + cc.tools.cc,
+            "-pgmP",
+            execroot + cc.tools.cc,
+            # Setting -pgm* flags explicitly has the unfortunate side effect
+            # of resetting any program flags in the GHC settings file. So we
+            # restore them here. See
+            # https://ghc.haskell.org/trac/ghc/ticket/7929.
+            "-optc-fno-stack-protector",
+            "-optP-E",
+            "-optP-undef",
+            "-optP-traditional",
+        ], format_each = "--ghc-option=%s")
+        args.add_all(["-v"], format_each = "--ghc-option=%s")
     args.add("--")
     args.add_all(package_databases, map_each = _dirname, format_each = "--package-db=%s")
     args.add_all(extra_include_dirs, map_each = _skip_ghc_bindist, format_each = "--extra-include-dirs=%s")
