diff --git a/haskell/private/cc_wrapper_windows.sh.tpl b/haskell/private/cc_wrapper_windows.sh.tpl
index d4bc6756..f4722411 100644
--- a/haskell/private/cc_wrapper_windows.sh.tpl
+++ b/haskell/private/cc_wrapper_windows.sh.tpl
@@ -101,8 +101,10 @@ unquote_arg() {
     output="$input"
 }
 
+DEBUG = "/c/users/admin/daml/cc_wrapper.log"
 add_arg() {
     quote_arg "$1" >> "$RESPONSE_FILE"
+    echo "ARG: $1" >> "$DEBUG"
 }
 
 # ----------------------------------------------------------
