diff --git a/haskell/private/actions/link.bzl b/haskell/private/actions/link.bzl
index f9e4a47..21d50aa 100644
--- a/haskell/private/actions/link.bzl
+++ b/haskell/private/actions/link.bzl
@@ -380,6 +380,7 @@ def link_library_dynamic(hs, cc, dep_info, cc_info, extra_srcs, objects_dir, my_
         with_profiling = False,
     )
 
+    args.add("-v")
     hs.toolchain.actions.run_ghc(
         hs,
         cc,
diff --git a/haskell/toolchain.bzl b/haskell/toolchain.bzl
index 2a453b8..bb8af04 100644
--- a/haskell/toolchain.bzl
+++ b/haskell/toolchain.bzl
@@ -80,6 +80,7 @@ while IFS= read -r line; do compile_flags+=("$line"); done < %s
 while IFS= read -r line; do extra_args+=("$line"); done < %s
 while IFS= read -r line; do param_file_args+=("$line"); done < %s
 
+echo "${compile_flags[@]}" "${extra_args[@]}" ${param_file_args+"${param_file_args[@]}"}
 "${compile_flags[@]}" "${extra_args[@]}" ${param_file_args+"${param_file_args[@]}"}
 """ % (compile_flags_file.path, extra_args_file.path, params_file_src)
 
