diff --git a/haskell/private/actions/link.bzl b/haskell/private/actions/link.bzl
index f9e4a47..21d50aa 100644
--- a/haskell/private/actions/link.bzl
+++ b/haskell/private/actions/link.bzl
@@ -380,6 +380,7 @@ def link_library_dynamic(hs, cc, dep_info, cc_info, extra_srcs, objects_dir, my_
         with_profiling = False,
     )
 
+    args.add("-v")
     hs.toolchain.actions.run_ghc(
         hs,
         cc,
