diff --git a/haskell-lsp.cabal b/haskell-lsp.cabal
index 75962ba..78076cb 100644
--- a/haskell-lsp.cabal
+++ b/haskell-lsp.cabal
@@ -35,7 +35,7 @@ library
   -- other-modules:
  -- other-extensions:
   ghc-options:         -Wall
-  build-depends:       base >= 4.9 && < 4.14
+  build-depends:       base >= 4.9 && < 4.15
                      , async
                      , aeson >=1.0.0.0
                      , attoparsec
@@ -67,7 +67,7 @@ executable lsp-hello
   default-language:    Haskell2010
   ghc-options:         -Wall
 
-  build-depends:       base >= 4.9 && < 4.14
+  build-depends:       base >= 4.9 && < 4.15
                      , aeson
                      , bytestring
                      , containers
