diff --git a/internal/pkg_npm/pkg_npm.bzl b/internal/pkg_npm/pkg_npm.bzl
index be857b3f..410a90b8 100644
--- a/internal/pkg_npm/pkg_npm.bzl
+++ b/internal/pkg_npm/pkg_npm.bzl
@@ -202,6 +202,7 @@ def create_package(ctx, deps_sources, nested_packages):
     if stamp:
         inputs.append(ctx.version_file)
 
+    posix = ctx.toolchains["@rules_sh//sh/posix:toolchain_type"]
     ctx.actions.run(
         progress_message = "Assembling npm package %s" % package_dir.short_path,
         mnemonic = "AssembleNpmPackage",
@@ -209,6 +210,9 @@ def create_package(ctx, deps_sources, nested_packages):
         inputs = inputs,
         outputs = [package_dir, ctx.outputs.pack, ctx.outputs.publish],
         arguments = [args],
+        env = {
+            "PATH": ctx.host_configuration.host_path_separator.join(posix.paths),
+        },
     )
     return package_dir
 
@@ -246,4 +250,5 @@ pkg_npm = rule(
     attrs = PKG_NPM_ATTRS,
     doc = _DOC,
     outputs = PKG_NPM_OUTPUTS,
+    toolchains = ["@rules_sh//sh/posix:toolchain_type"],
 )
